chore: add documentation on query parameter with priority (#707)
SAP/cloud-sdk: docs/js/features/connectivity/query-parameters.mdx

---
title: Query Parameters
hide_title: false
hide_table_of_contents: false
sidebar_label: Query Parameters
description: This article describes how query parameters are prioritized and handled by the SAP Cloud SDK.
keywords:
  - sap
  - cloud
  - sdk
  - parameters
  - encoding
  - connectivity
  - JavaScript
  - TypeScript
---

import LinkToLatestJsApiDocs from '../../../../src/sap/sdk-js/LinkToLatestJsApiDocs';
import useBaseUrl from '@docusaurus/useBaseUrl';
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

- Version 2.0 disclaimer

> In this document we explain where how query parameters

Reviewer: What?

## Header Origins and Priority

> As partially described in the [destination guide](./destination#additional-headers-and-query-parameters-on-destinations) there are multiple origins for query parameters:

Reviewer 2: ```suggestion
As partially described in the [destination guide](./destination#additional-headers-and-query-parameters-on-destinations), there are multiple possible origins for query parameters:
```

> - Provided by user on request level.

Reviewer: ```suggestion
- Provided by the user on request level.
```

  These parameters are referred to as `custom` parameters.
- Defined by the destination object or properties.
- Introduced by the SAP Cloud SDK.
  These parameters are referred to as `internal` parameters.

In case there are parameters with the same key from multiple origins the priority is:

1. `Custom` parameters
2. `Destination` parameters
3. `Internal` parameters

### Custom Parameters

Custom parameters are set on the request level by the `execute()` method of the generated clients

<Tabs groupId="version" defaultValue="major" values={[ { label: 'SDK 2.x', value: 'major' }, { label: 'SDK 1.x', value: 'deprecated' } ]}>
<TabItem value="major">

```ts
const { myEntityApi } = myEntityService();
myEntityApi.requestBuilder().getAll().addCustomQueryParameters({
  apikey: 'my-api-key'
});
```

</TabItem>
<TabItem value="deprecated">

```ts
MyEntity.requestBuilder().getAll().addCustomQueryParameters({
  apikey: 'my-api-key'
});
```

</TabItem>
</Tabs>

> or the `executeHttpRequest` method of the generic http-client:

Reviewer 2: ```suggestion
or the `executeHttpRequest` method of the generic `http-client`:
```

<Tabs groupId="version" defaultValue="major" values={[ { label: 'SDK 2.x', value: 'major' }, { label: 'SDK 1.x', value: 'deprecated' } ]}>
<TabItem value="major">

```ts
const requestConfig = {
  params: {
    custom: { parameterKey: parameterValue }
  }
};
executeHttpRequest(myDestination, requestConfig);
```

</TabItem>
<TabItem value="deprecated">

```ts
const requestConfig = {
  params: {
    parameterKey: parameterValue
  }
};
executeHttpRequest(myDestination, requestConfig);
```

</TabItem>
</Tabs>

> Note the difference between version 1.0 and 2.0 of the SAP Cloud SDK.

Reviewer 2: ```suggestion
Note the difference between version 1 and 2 of the SAP Cloud SDK.
```

I think we should not use `X.0` when referring to these versions as this implies it is only for that specific minor release.

Reviewer 2: Do you know if we have a docs style guide where we could mention this?

Author: I agree and I will try to remember not to use the `.` versions. I do not know if there is a style guide...

> In the newer version we allow to specify the origin of the header, whereas for version 1.0 it is `custom` per default.

Reviewer 2: ```suggestion
In version 2, we allow to specify the origin of the header. 
For version 1, it is always `custom`.
```

```ts
const { queryParameters } = await getDestination({
  destinationName: 'myDestination',
  userJwt: 'yourJwt'
});
```

### Internal Parameters

The SAP Cloud SDK sets certain parameters in the query per default.
For OData the format is set to `json` for example using a query parameter.
Also all chosen options like filter, select, or top in the OData clients lead to query parameters considered to be `internal`:

<Tabs groupId="version" defaultValue="major" values={[ { label: 'SDK 2.x', value: 'major' }, { label: 'SDK 1.x', value: 'deprecated' } ]}>
<TabItem value="major">

```ts
const { myEntityApi } = myEntityService();
myEntityApi.requestBuilder().getAll().
    .select(myEntityApi.schema.FIRST_NAME)
    .top(5)
// request url: /myEntityApi?$select=FirstName&$top=5&$format=json
```

</TabItem>
<TabItem value="deprecated">

```ts
MyEntity.requestBuilder().getAll().select(myEntityApi.schema.FIRST_NAME).top(5);
// request url: /myEntityApi?$select=FirstName&$top=5&$format=json
```

</TabItem>
</Tabs>

The internal parameters have the lowest priority and you should not have to adjust them manually.

## Encoding

:::info
This section applies to version 2.0 of the SAP Cloud SDK.

> For version 1.0 encoding is done one OData related query parameters without a way to configure it.

Reviewer: ```suggestion
For version 1.0 encoding is done on OData related query parameters without a way to configure it.
```

:::

In the end, query parameters end up in the URL to the target system.
Therefore special characters with a meaning in a URL need to be [% encoded](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier).
The SAP Cloud SDK takes care of this encoding depending on the origin you use.
For calls executed via a typed client and generic http client the encoding is done as:

- `Custom` parameters are not encoded.
  The idea behind this is to give you a way to send parameters with highest priority to the target system as they are.
- `Destination` parameters are encoded.
- `Internal` parameters are encoded.

For the `executeHttpRequest` method a option is available to adjust the behavior.

```ts
const parameterEncoder = function (
  params: Record<string, any>
): Record<string, any> {
  // your custom implementation
};
executeHttpRequest(destination, { parameterEncoder });
```

> The `parameterEncoder function is applied to the keys and values of all parameters of the request. In case you have the need to encode all parameters also the custom ones, we exported a `encodeAllParameters`method to do that. Simply pass this method as the`parameterEncoder`to the`executeHttpRequest` call.

Reviewer 2: ```suggestion
The `parameterEncoder` function is applied to the keys and values of all parameters of the request. 
In case you have the need to encode all parameters including custom ones, we exported a `encodeAllParameters` method to do that. 
Simply pass this method as the `parameterEncoder` to the `executeHttpRequest` call.
```
